src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java

```diff
@@ -15,135 +15,199 @@
     }
+
+    this.client = client != null ? client : new ElasticsearchClient(config, reporter);
+
+    log.info("Started ElasticsearchSinkTask.");
   }
 
   @Override
   public void put(Collection<SinkRecord> records) throws ConnectException {
-    log.debug("Putting {} records to Elasticsearch", records.size());
-    writer.write(records);
+    log.debug("Putting {} records to Elasticsearch.", records.size());
+    for (SinkRecord record : records) {
+      if (shouldSkipRecord(record)) {
+        logTrace("Ignoring {} with null value.", record);
+        reportBadRecord(record, new ConnectException("Cannot write null valued record."));
+        continue;
+      }
+
+      logTrace("Writing {} to Elasticsearch", record);
+
+      String index = convertTopicToIndexName(record.topic());
+      ensureIndexExists(index);
+      checkMapping(record);
+      tryWriteRecord(record);
+    }
   }
 
   @Override
   public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
-    if (writer != null) {
-      log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
-      writer.flush();
-    } else {
-      log.debug("Could not flush data to Elasticsearch because ESWriter already closed.");
+    log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
+    try {
+      client.flush();
+    } catch (IllegalStateException e) {
+      log.debug("Tried to flush data to Elasticsearch, but BulkProcessor is already closed.");
     }
   }
 
   @Override
-  public void close(Collection<TopicPartition> partitions) {
-    log.debug("Closing the task for topic partitions: {}", partitions);
+  public void stop() {
+    log.debug("Stopping Elasticsearch client.");
+    client.close();
+    log.debug("Stopped Elasticsearch client.");
   }
 
   @Override
-  public void stop() throws ConnectException {
-    log.info("Stopping ElasticsearchSinkTask");
-    if (writer != null) {
-      writer.stop();
-      writer = null;
+  public String version() {
+    return Version.getVersion();
+  }
+
+  private void checkMapping(SinkRecord record) {
+    String index = convertTopicToIndexName(record.topic());
+    if (!config.shouldIgnoreSchema(record.topic()) && !existingMappings.contains(index)) {
+      if (!client.hasMapping(index)) {
+        client.createMapping(index, record.valueSchema());
+      }
+      log.debug("Locally caching mapping for index '{}'", index);
+      existingMappings.add(index);
+    }
+  }
+
+  /**
+   * Returns the converted index name from a given topic name. Elasticsearch accepts:
+   * <ul>
+   *   <li>all lowercase</li>
+   *   <li>less than 256 bytes</li>
+   *   <li>does not start with - or _</li>
+   *   <li>is not . or ..</li>
+   * </ul>
+   * (<a href="https://github.com/elastic/elasticsearch/issues/29420">ref</a>_.)
+   */
+  private String convertTopicToIndexName(String topic) {
+    String index = topic.toLowerCase();
+    if (index.length() > 255) {
```

Comment:
Happy coincidence that Kafka topic names are more limited and accept only single byte characters. But if multi-byte were accepted this wouldn't necessarily be correct. 

Should we link to the docs instead of a github issue?
E.g.: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html#indices-create-api-path-params